[dynamo][guards-c++-refactor] Introduce LeafGuard, GuardManager and GuardAccessor classes

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120123
* #120119
* #120098
* #120096
* #120093
* #120091
* #120089
* #120068
* #120067
* #120065
* #120064
* #120062
* #120061
* #120060
* #119833
* #119827
* __->__ #119822


The full blown implementation is in this stack - https://github.com/pytorch/pytorch/pull/110590 which is passing all the test cases on CI. That stack is hard to review. So, breaking apart.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

anijain2305 committed:
[dynamo][guards-c++-refactor] Introduce LeafGuard, GuardManager and GuardAccessor classes

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+
+ private:
+  // id of the type of the original object.
+  unsigned long long _expected;
```
intptr_t -- don't assume 64 bit ptrs.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+
+ private:
+  // id of the original object.
+  unsigned long long _expected;
```
intptr_t

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+
+  // reset the relational guard state on guard failure. This is called by the
+  // guard manager.
+  virtual void reset_state() = 0;
```
This won't be thread safe, so need to ensure we have a lock around guard checking.

anijain2305 replied:
Wow! Thats a great catch.

Just so that I understand - This is not thread safe because two threads can try to run the same code object. So two threads might be running same guard function, and they can try to change the state simultaneously.

So, this is not just `reset_state`. This is bad even for `sorting` of guard managers where two threads might try to sort at the same time, and can leave the array in a bad state causing segfaults.

Added a std::lock_guard at the beginning of `check_nopybind` and `check_verbose_nopybind`. Please let me know if thats sufficient.

jansel replied:
The lock should be in the root guard manager.

anijain2305 replied:
Yes, this is in RootGuardManager's `check_nopybind` and `check_verbose_nopybind`

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+    return _guard_manager;
+  }
+
+  bool matches_key(const py::object key) const {
```
calling this will construct a new py::object and increment/decrement refcount

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+    bool result = true;
+    // Iterate over leaf guards
+    for (const auto& guard : _leaf_guards) {
+      result = result && guard->check_nopybind(value);
+      if (!result) { // early exit
```
```suggestion
      if (!guard->check_nopybind(value)) { // early exit
```

`result &&` is a no-op.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+      result = result && guard->check_nopybind(value);
+      if (!result) { // early exit
+        _fail_count += 1;
+        return result;
```
```suggestion
        return false;
```

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+    // Iterate over accessors.
+    bool failed_on_first = true;
+    for (const auto& accessor : _accessors) {
+      result = result && accessor->check_nopybind(value);
```
```suggestion
      result = accessor->check_nopybind(value);
```

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+ protected:
+  // Keeps a count of how many times this guard manager check function returns
+  // False. This is used for sorting optimization.
+  int _fail_count{0};
```
use int64_t -- ints can overflow

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+    _relational_guard_resetters.emplace_back(std::move(relational_guard));
+  }
+
+  // Python visible API to check guard function.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  // Python visible API to check_verbose guard function.
+  GuardDebugInfo check_verbose(py::handle value) {
+    return check_verbose_nopybind(value.ptr());
+  }
```
weren't these already declared in base class?

anijain2305 replied:
No. We only want these to for `RootGuardManager`. We declared for `LeafGuard` but thats a separate data structure.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+
+    // Iterate over epilogue leaf guards.
+    for (const auto& guard : _epilogue_lambda_guards) {
+      result = result && guard->check_nopybind(value);
```
```suggestion
      result = guard->check_nopybind(value);
```
not reachable

anijain2305 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -664,6 +664,615 @@ static struct PyModuleDef _module = {
+ private:
+  // This is set while constructing the leaf guard. This is used for identifying
+  // the cause of recompilation.
+  py::list _verbose_code_parts;
```
Its a list of strings. Usage is something like

```
guard = guards.TYPE_MATCH(id_type(foo), ["type(x) == int"])
```

Its a list because we can have multiple `verbose_code_parts` for to the same guard. So, I use `py::object` to hold a reference for the list of strings.

anijain2305 committed:
Update on "[dynamo][guards-c++-refactor] Introduce LeafGuard, GuardManager and GuardAccessor classes"



The full blown implementation is in this stack - https://github.com/pytorch/pytorch/pull/110590 which is passing all the test cases on CI. That stack is hard to review. So, breaking apart.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards-c++-refactor] Introduce LeafGuard, GuardManager and GuardAccessor classes"



The full blown implementation is in this stack - https://github.com/pytorch/pytorch/pull/110590 which is passing all the test cases on CI. That stack is hard to review. So, breaking apart.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented:
@jansel This is ready for another review.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    PyObject* x = PyObject_CallOneArg(_guard_check_fn.ptr(), value); // new ref
+    if (x == nullptr) {
+      // An exception is caught in the lambda function.
```
If this is supposed to be unreachable should we assert rather than silently returning?

If we want to return you need to clear the Python error.

anijain2305 replied:
Clearing the expection and returning False. I don't want to assert because otherwise then we might not DECREF some other pointer held higher up in the GuardManager tree.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+  // type_id = id(type(obj))
+  TYPE_MATCH(py::object type_id, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts),
+        _expected(py::cast<unsigned long long>(type_id)) {}
```
```suggestion
        _expected(py::cast<decltype(_expected)>(type_id)) {}
```
or intptr_t

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+  // obj_id = id(obj)
+  ID_MATCH(py::object obj_id, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts),
+        _expected(py::cast<unsigned long long>(obj_id)) {}
```
same

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+  // Fast check function.
+  virtual bool check_nopybind(PyObject* value) override { // borrowed ref
+    // reset_state is not thread safe. So acquire a lock.
+    static std::mutex lock;
```
This lock:
1) Needs to be shared with check_verbose -- currently we have two locks
2) Can be a class member, since it is fine to run different guard checks in parallel
3) Need to think about interactions with the GIL.  Might need to add:
```
Py_BEGIN_ALLOW_THREADS
std::lock_guard<std::mutex> lock_guard(lock);
Py_END_ALLOW_THREADS
```
That way if this thread blocks waiting we aren't holding the GIL.  Without this I think we could deadlock if you hold the GIL while waiting for the lock.

anijain2305 replied:
1. Made the mutex a class member.
2. I read about the `Py_BEGIN_ALLOW_THREADS` and your solution makes sense to me. I have not worked with GIL before, so I am not sure if there are other things I need to worry about. I added your suggestion here. (I had to add ; at the end of macros, because clang format was not working, even after disabling them)

anijain2305 replied:
On second though, using `Py_BEGIN/END_ALLOW_THREADS` might be wrong because they will put lock_guard in braces {}, and lock_guard will be deleted even before we call the critical part. Maybe we should use `Py_BLOCK_THREADS` and ` Py_UNBLOCK_THREADS`

jansel replied:
Yeah, I didn't look at what the macros do.  What you want to do is release the GIL while waiting for the lock.

anijain2305 replied:
Ok, I am convinced that the deadlock you mention can happen. I added a note as well to explain how the deadlock could happen. And yes, releasing and acquiring GIL before the `lock_guard` will avoid the deadlock.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+  virtual GuardDebugInfo check_verbose_nopybind(
+      PyObject* value) override { // borrowed ref
+    // reset_state is not thread safe. So acquire a lock.
+    static std::mutex lock;
```
same

anijain2305 committed:
Update on "[dynamo][guards-c++-refactor] Introduce LeafGuard, GuardManager and GuardAccessor classes"



The full blown implementation is in this stack - https://github.com/pytorch/pytorch/pull/110590 which is passing all the test cases on CI. That stack is hard to review. So, breaking apart.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards-c++-refactor] Introduce LeafGuard, GuardManager and GuardAccessor classes"



The full blown implementation is in this stack - https://github.com/pytorch/pytorch/pull/110590 which is passing all the test cases on CI. That stack is hard to review. So, breaking apart.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented:
@jansel Ready for re-review.

anijain2305 commented:
@pytorchbot label "topic: not user facing"
